### PR TITLE
feat(CNAD): add a new resource cnad advanced policy

### PR DESCRIPTION
**Pull request by deer-hang**
**What this PR does / why we need it**:
add a new resource cnad advanced policy

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

This PR depends on #3095

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/cnad' TESTARGS='-run TestAccCNADAdvancedPolicy_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/cnad -v -run TestAccCNADAdvancedPolicy_basic -timeout 360m -parallel 4 
=== RUN   TestAccCNADAdvancedPolicy_basic 
=== PAUSE TestAccCNADAdvancedPolicy_basic
=== CONT  TestAccCNADAdvancedPolicy_basic
--- PASS: TestAccCNADAdvancedPolicy_basic (69.85s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cnad      69.921s
```

### CODE REVIEW

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cnad' TESTARGS='-run TestAccPolicy_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/cnad -v -run TestAccPolicy_basic -timeout 360m -parallel 4 
=== RUN   TestAccPolicy_basic 
=== PAUSE TestAccPolicy_basic
=== CONT  TestAccPolicy_basic
--- PASS: TestAccPolicy_basic (20.81s) 
PASS 
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cnad      20.907s
```

File: docs/resources/cnad_advanced_policy.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+## Example Usage
+
+```hcl
+data "huaweicloud_cnad_advanced_instances" "test" {}
+
+resource "huaweicloud_cnad_advanced_policy" "test" {
+  instance_id = huaweicloud_cnad_advanced_instances.test.instances[0].instance_id
+  name        = "test-policy"
+  threshold   = 100
+  udp         = block
```
Comment (houpeng80) [inline docs/resources/cnad_advanced_policy.md]:
```suggestion
  udp         = "block"
```

File: docs/resources/cnad_advanced_policy.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `instance_id` - (Required, String, ForceNew) Specifies the instance id.
+
+  Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the policy name, The maximum length is 255 characters.
```
Comment (houpeng80) [inline docs/resources/cnad_advanced_policy.md]:
```suggestion
* `name` - (Required, String) Specifies the policy name, the maximum length is 255 characters.
```

File: docs/resources/cnad_advanced_policy.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+* `threshold` - (Optional, Int) Specifies the cleaning threshold, The value ranges from 100 to 1000.
```
Comment (houpeng80) [inline docs/resources/cnad_advanced_policy.md]: ditto

File: docs/resources/cnad_advanced_policy.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+* `connection_protection_list` - The connection protection list.
+
+* `fingerprint_count` - The fingerprint count.
+
+* `port_block_count` - The number of port blockages.
+
+* `watermark_count` - The number of watermarks.
+
+* `if_exist_traffic` - Is there any traffic.
```
Comment (houpeng80) [inline docs/resources/cnad_advanced_policy.md]: it is better to exist_traffic, and the description change to "Whether traffic exists"

File: docs/resources/cnad_advanced_policy.md (hunk 1 above)
Comment (deer-hang) [reply docs/resources/cnad_advanced_policy.md]: OK

File: docs/resources/cnad_advanced_policy.md (hunk 2 above)
Comment (deer-hang) [reply docs/resources/cnad_advanced_policy.md]: OK

File: docs/resources/cnad_advanced_policy.md (hunk 3 above)
Comment (deer-hang) [reply docs/resources/cnad_advanced_policy.md]: OK

File: docs/resources/cnad_advanced_policy.md (hunk 4 above)
Comment (deer-hang) [reply docs/resources/cnad_advanced_policy.md]: This field is used for console, we can remove it.

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cnad' TESTARGS='-run TestAccPolicy_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/cnad -v -run TestAccPolicy_basic -timeout 360m -parallel 4 
=== RUN   TestAccPolicy_basic 
=== PAUSE TestAccPolicy_basic
=== CONT  TestAccPolicy_basic
--- PASS: TestAccPolicy_basic (23.45s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cnad      23.512s
```

File: huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+			"instance_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `Specifies the instance id.`,
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `Specifies the policy name, The maximum length is 255 characters.`,
```
Comment (ShiChangkuo) [inline huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go]: the issue should also be updated in `Description`

File: huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+		MoreHeaders: map[string]string{
+			"Content-Type": "application/json;charset=utf8",
+		},
+		KeepResponseBody: true,
+		JSONBody:         utils.RemoveNil(buildCreatePolicyBodyParams(d)),
+	}
+
+	createPolicyResp, err := createPolicyClient.Request("POST", createPolicyPath, &createPolicyOpt)
+	if err != nil {
+		return diag.Errorf("error creating CNADAdvancedPolicy: %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go]:
```suggestion
		return diag.Errorf("error creating CNAD advanced policy: %s", err)
```

File: huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+	}
+
+	if err := readPolicyDetail(getPolicyClient, d); err != nil {
+		return common.CheckDeletedDiag(d, err, "error retrieving CNADAdvancedPolicy")
+	}
+
+	return nil
+}
+
+func readPolicyDetail(getPolicyClient *golangsdk.ServiceClient, d *schema.ResourceData) error {
```
Comment (ShiChangkuo) [inline huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go]: this function will query the policy from HW cloud and set fields to state, it's better to ensure the function only responsible for one thing.

File: docs/resources/cnad_advanced_policy.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `instance_id` - (Required, String, ForceNew) Specifies the CNAD advanced instance ID.
+  You can find it through data source `huaweicloud_cnad_advanced_instances`. If you have no instances, please
+  buy one first.
```
Comment (ShiChangkuo) [inline docs/resources/cnad_advanced_policy.md]: `If you have no instances, please buy one first.` --- it's unnecessary to add it.

File: docs/resources/cnad_advanced_policy.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+  Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the policy name, the maximum length is 255 characters.
+
+* `udp` - (Optional, String) Specifies whether to block the udp protocol. Valid values are **block** and **unblock**.
+
+* `threshold` - (Optional, Int) Specifies the cleaning threshold, the value ranges from 100 to 1000.
```
Comment (ShiChangkuo) [inline docs/resources/cnad_advanced_policy.md]: what's the unit of threshold?

File: docs/resources/cnad_advanced_policy.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+* `instance_id` - (Required, String, ForceNew) Specifies the CNAD advanced instance ID.
+  You can find it through data source `huaweicloud_cnad_advanced_instances`. If you have no instances, please
+  buy one first.
+
+  Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the policy name, the maximum length is 255 characters.
+
+* `udp` - (Optional, String) Specifies whether to block the udp protocol. Valid values are **block** and **unblock**.
```
Comment (ShiChangkuo) [inline docs/resources/cnad_advanced_policy.md]:
```suggestion
* `udp` - (Optional, String) Specifies whether to block the UDP protocol. Valid values are **block** and **unblock**.
```

File: huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go (hunk 5 above)
Comment (deer-hang) [reply huaweicloud/services/cnad/resource_huaweicloud_cnad_advanced_policy.go]: OK

File: docs/resources/cnad_advanced_policy.md (hunk 9 above)
Comment (deer-hang) [reply docs/resources/cnad_advanced_policy.md]: Mbps

File: docs/resources/cnad_advanced_policy.md (hunk 10 above)
Comment (deer-hang) [reply docs/resources/cnad_advanced_policy.md]: OK

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/cnad' TESTARGS='-run TestAccPolicy_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/cnad -v -run TestAccPolicy_basic -timeout 360m -parallel 4 
=== RUN   TestAccPolicy_basic 
=== PAUSE TestAccPolicy_basic
=== CONT  TestAccPolicy_basic
--- PASS: TestAccPolicy_basic (21.82s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cnad      21.880s
```

**Pull request merged**